app/com/linkedin/drelephant/tuning/engine/SparkHBTParamRecommender.java

```diff
@@ -0,0 +1,361 @@
+      long currSuggestedMemory = suggestExecutorMemory(maxPeakJVMUsedMemoryPerCore, core);
+      if (currSuggestedMemory < MAX_EXECUTOR_MEMORY || core == 1) {
+        suggestedExecutorMemory = currSuggestedMemory;
+        suggestedCore = core;
+        validSuggestion = true;
+        break;
+      }
+    }
+    if (!validSuggestion) {
+      suggestedExecutorMemory = lastRunExecutorMemory;
+      suggestedCore = lastRunExecutorCore;
+    }
+    if (getMemoryIncreaseForGC() != 0) {
+      suggestedExecutorMemory = suggestedExecutorMemory * (100 + getMemoryIncreaseForGC()) / 100;
+    }
+    if (suggestedExecutorMemory < MIN_EXECUTOR_MEMORY) {
+      suggestedExecutorMemory = MIN_EXECUTOR_MEMORY;
+    }
+    suggestedExecutorMemory = getRoundedExecutorMemory();
+  }
+
+  /**
+   * This method is to get the rounded executor memory. We don't want to round executor memory to
+   * multiple of 1GB as container size is decided by executor memory + memoryOverhead. So round the
+   * executor memory + memoryOverhead and then distribute extra memory to executor memory
+   *
+   * @return rounded executor memory
+   */
+  private long getRoundedExecutorMemory() {
+    long memoryOverhead = lastRunExecutorMemoryOverhead;
+    if (lastRunExecutorMemoryOverhead == 0) {
+      memoryOverhead = Math.max(suggestedExecutorMemory / 10, 384);
+    }
+
+    long roundedContainerSize = getRoundedContainerSize(suggestedExecutorMemory + memoryOverhead + FileUtils.ONE_MB);
+    long executorMemory =
+        (long) (suggestedExecutorMemory + (roundedContainerSize - suggestedExecutorMemory - memoryOverhead - FileUtils.ONE_MB) * 0.9);
+
+    return executorMemory;
+  }
+
+  /**
+   * This method is to get the rounded executor memory. We don't want to round executor memory to
+   * multiple of 1GB as container size is decided by executor memory + memoryOverhead. So round the
+   * executor memory + memoryOverhead and then distribute extra memory to executor memory
+   *
+   * @return rounded driver memory
+   */
+  private long getRoundedDriverMemory() {
+    long memoryOverhead = lastRunDriverMemoryOverhead;
+    if (lastRunDriverMemoryOverhead == 0) {
+      memoryOverhead = Math.max(suggestedDriverMemory / 10, 384);
+    }
+
+    long roundedContainerSize = getRoundedContainerSize(suggestedDriverMemory + memoryOverhead + FileUtils.ONE_MB);
+    long driverMemory =
+        (long) (suggestedDriverMemory + (roundedContainerSize - suggestedDriverMemory - memoryOverhead - FileUtils.ONE_MB) * 0.9);
+
+    return driverMemory;
+  }
+
+  /**
+   * Suggest executor memory based on peak JVM used memory per core and number of core after adding
+   * buffer
+   *
+   * @param maxPeakJVMUsedMemoryPerCore
+   * @param core
+   * @return suggested executor memory
+   */
+  private long suggestExecutorMemory(long maxPeakJVMUsedMemoryPerCore, int core) {
+    return (maxPeakJVMUsedMemoryPerCore * (100 + EXECUTOR_MEMORY_BUFFER_PER_CORE) / 100) * core
```

Comment:
Right now both EXECUTOR_MEMORY_BUFFER_PER_CORE and EXECUTOR_MEMORY_BUFFER_OVERALL are set to 0. I'd recommend some buffer, due to to fluctuations in data set sizes for different days. In the future, it might be nice to look at variance in previous runs, to try to calculate a good buffer size.

Reply:
For now I have kept it to 0 based on initial tests on around 10 jobs. I think let's run it for few jobs and see how it behaves. Based on that we can change it and provide more buffer.